### PR TITLE
DRILL-8360: Add Provided Schema for XML Reader

### PR DESCRIPTION
**Pull request by cgivre**
# [DRILL-8360](https://issues.apache.org/jira/browse/DRILL-8360): Add Provided Schema for XML Reader

## Description
This PR adds several enhancements to XML functionality:

* Allows for provided schema for XML files
* Allows for provided schema for XML API requests
* Allows for XML to be used in post bodies.

This PR deprecates the `xmlDataLevel` in the HTTP plugin in favor of a new `xmlOptions` class which has the data level.  It does not remove the `xmlDataLevel` as that would be a breaking change.
 
Also to note, at this time, the provided schema functionality only works with scalar data types.  In the future we can add complex type support such as lists and maps.

## Documentation
Updated appropriate README files.

## Testing
Added various unit tests.

### CODE REVIEW

File: contrib/format-xml/src/main/java/org/apache/drill/exec/store/xml/XMLReader.java [hunk 1]
```diff
@@ -428,8 +435,67 @@ private void writeFieldData(String fieldName, String fieldValue, TupleWriter wri
...
+          dateFormat = columnMetadata.property("drill.format");
+          Instant timestamp = null;
+          if (Strings.isNullOrEmpty(dateFormat)) {
+            timestamp = Instant.parse(fieldValue);
+          } else {
+            try {
+              SimpleDateFormat simpleDateFormat = new SimpleDateFormat(dateFormat);
+              Date parsedDate = simpleDateFormat.parse(fieldValue);
+              timestamp = Instant.ofEpochMilli(parsedDate.getTime());
+            } catch (ParseException e) {
```
Comment (jnturton) [inline contrib/format-xml/src/main/java/org/apache/drill/exec/store/xml/XMLReader.java:490]: I think we should operate in one of two definite modes controlled by an option that we try to standardise across plugins.

1. Invalid data fails hard and fast. Your results will not be silently distorted.
2. Invalid data is swallowed and null is emitted. You know what you're doing and you'll provide appropriate logic for the nulls later.

Users that want a default substituted have the simple `ifnull(x,-1)` to combine with (2). My opinion is that if we don't have a `nullInvalidData: true` option set then we should continue to fail the query on the spot, as we do elsewhere.

**Review by jnturton — CHANGES_REQUESTED** [review]

File: contrib/format-xml/src/main/java/org/apache/drill/exec/store/xml/XMLReader.java (hunk 1 above)
Comment (jnturton) [reply contrib/format-xml/src/main/java/org/apache/drill/exec/store/xml/XMLReader.java:490]: Maybe this option is better as a global that is overridable at the session level come to think of it. But until we have such an option I think the default behaviour so far is that we fail on invalid data and we shouldn't muddy things up.

Comment (cgivre) [reply contrib/format-xml/src/main/java/org/apache/drill/exec/store/xml/XMLReader.java:490]: @jnturton In principle I agree.  The code I used for date parsing was borrowed from elsewhere in Drill, so at least the behavior is consistent if not ideal.  Should I create another JIRA to introduce a global/session option for behavior on date/time casts?

**Comment by cgivre** [comment]
It seems like we're getting random failures in the Splunk unit tests as well.

File: contrib/format-xml/src/main/java/org/apache/drill/exec/store/xml/XMLReader.java (hunk 1 above)
Comment (jnturton) [reply contrib/format-xml/src/main/java/org/apache/drill/exec/store/xml/XMLReader.java:490]: Not only dates and times but numerics and strings when invalid UTF-8 sequences are encountered! I created [DRILL-8361](https://issues.apache.org/jira/browse/DRILL-8361) for this.

I see we slipped up in the PDF plugin and we have two places where invalid data swallowing is present. Let's make both of them (XML and PDF) consistent with the rest of Drill (do not swallow invalid data) in this PR, and do the nulling of invalid data properly in 8361?

**Comment by jnturton** [comment]
> It seems like we're getting random failures in the Splunk unit tests as well.

Were you able to capture any output? The most recent run didn't reveal anything.

**Review by jnturton — APPROVED** [review]
Thanks for making changes.

**Pull request merged**